pallets/marketplace/src/lib.rs

```diff
@@ -0,0 +1,512 @@
+            }
+
+            Self::deposit_event(Event::AdvertisementStored(advertisement, who));
+            Ok(().into())
+        }
+
+        /// Delete advertisement.
+        #[pallet::weight(< T as Config >::WeightInfo::delete_advertisement())]
+        pub fn delete_advertisement(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+
+            let ad = <StoredAdvertisement<T>>::get(who.clone())
+                .ok_or(Error::<T>::AdvertisementNotFound)?;
+
+            // update index
+            for pricing in &ad.pricing {
+                <StoredAdIndex<T>>::mutate(&pricing.reward_asset, |ads| {
+                    let mut a = ads.clone().unwrap_or_default();
+                    a.retain(|v| v.0 != who.clone())
+                });
+            }
+
+            <StoredAdvertisement<T>>::remove(who.clone());
+            <StoredCapacity<T>>::remove(who.clone());
+
+            Self::deposit_event(Event::AdvertisementRemoved(who));
+            Ok(().into())
+        }
+    }
+
+    impl<T: Config> From<Error<T>> for pallet_acurast::Error<T> {
+        fn from(_: Error<T>) -> Self {
+            Self::JobHookFailed
+        }
+    }
+
+    impl<T: Config> JobHooks<T> for Pallet<T> {
+        type Error = Error<T>;
+
+        /// Registers a job in the marketplace by providing a [JobRegistration].
+        /// If a job for the same `(accountId, script)` was previously registered, it will be overwritten.
+        fn register_hook(
+            who: &T::AccountId,
+            registration: &JobRegistrationFor<T>,
+        ) -> Result<(), DispatchError> {
+            let e: <T as Config>::RegistrationExtra = registration.extra.clone().into();
+            let extra: JobRequirements = e.into();
+
+            ensure!(
+                extra.cpu_milliseconds > 0,
+                Error::<T>::JobRegistrationZeroCPUMilliseconds
+            );
+            ensure!(extra.slots > 0, Error::<T>::JobRegistrationZeroSlots);
+            let reward_amount: T::AssetAmount = registration
+                .reward
+                .try_get_amount()
+                .map_err(|_| Error::<T>::JobRegistrationUnsupportedReward)?
+                .into();
+            ensure!(
+                reward_amount > 0.into(),
+                Error::<T>::JobRegistrationZeroReward
+            );
+
+            // reward is understood per slot
+            let mut total = registration.reward.clone();
+            total
+                .with_amount(
+                    reward_amount
+                        .checked_mul(&((extra.slots as u128).into()))
+                        .ok_or(Error::<T>::RewardCalculationOverflow)?
+                        .into(),
+                )
+                .map_err(|_| Error::<T>::RewardConversionFailed)?;
+            T::RewardManager::lock_reward(total.clone(), T::Lookup::unlookup(who.clone()))
+                .map_err(|_| Error::<T>::InvalidPayment)?;
```

Comment:
@gitsimon `T::RewardManager::lock_reward` is currently still called in the register extrinsic in `pallet_acurast` before the `register_hook` is called. So it is being called twice now.

Reply:
RewardManager moved to marketplace